tests/sharded_jit_test.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from functools import partial
+from unittest import SkipTest
+
+import numpy as onp
+from absl.testing import absltest
+
+import jax
+import jax.numpy as np
+from jax import test_util as jtu
+from jax.interpreters import pxla
+from jax.interpreters.sharded_jit import sharded_jit
+from jax.interpreters.sharded_jit import PartitionSpec as P
+
+from jax.config import config
+config.parse_flags_with_absl()
+
+
+class ShardedJitTest(jtu.JaxTestCase):
+
+  def testBasic(self):
+    if jtu.device_under_test() != "tpu" or jax.device_count() < 2:
+      raise SkipTest
+
+    @partial(sharded_jit, in_parts=(P(2, 1), P(2, 1)), out_parts=None)
+    def f(x, y):
+      return x + y
+
+    shape = (8, 8)
+    x = onp.arange(onp.prod(shape), dtype=np.float32).reshape(shape)
+    actual = f(x, x + 1)
+    expected = x + (x + 1)
+    self.assertAllClose(actual, expected, check_dtypes=False)
+    self.assertIsInstance(actual, pxla.ShardedDeviceArray)
+    self.assertLen(actual.device_buffers, 2)
+    self.assertAllClose(actual.device_buffers[0].to_py(), expected,
+                        check_dtypes=False)
+
+  def testPyTreeArgs(self):
+    if jtu.device_under_test() != "tpu" or jax.device_count() < 2:
+      raise SkipTest
+
+    def f(a, b, c):
+      a1, a2 = a
+      c1, (c2, c3) = c
+      return a1 + a2 + b + c1 + c2 + c3
+
+    def _make_arg(*shape):
+      return onp.arange(onp.prod(shape)).reshape(shape)
+
+    a = (_make_arg(4, 4), _make_arg())
+    b = _make_arg(4, 4)
+    c = [_make_arg(), (_make_arg(4, 4), _make_arg(4, 4))]
+
+    in_parts = (None, P(2, 1), [None, P(2, 1)])
```

Comment:
I do not quite understand how to make the `in_parts` for `c` to the structure of `c`, specifically the second element of a list: it is a pair, but it gets only a single sharding specification `P(2, 1)`. Wouldn't it be simpler to request
that the `in_parts` matches the structure of the argument?

Reply:
I'm reusing the vmap machinery (`api_util.flatten_axes`, which I just realized I should rename now :)). I think it's useful to allow a pytree prefix, rather than the exact pytree structure, in the case where you have a complicated pytree (e.g. a flax model) and want to partition or replicate the whole thing.

Relatedly, I think it's gonna be hard if you do have something like a flax model, and want to partition just one part of it, or similar vmap just one part of it. A problem for another day though.

Reply:
Yeah, I think fundamentally that's the "API for transforming parameterized/stateful functions" problem and we'll need to solve it somewhere higher.

Reply:
Let's add some documentation somewhere, e.g., in the FAQ, or the pytree page, with a notes from the various API docstrings.

Reply:
I filed https://github.com/google/jax/issues/3095 to keep track of this. I'd prefer not to hold up this PR on adding + reviewing the new docs, plus it should probably be its own PR anyway. I can take that issue once I get the rest of my sharded_jit PRs are ready, but I left it unassigned in the meantime.